File: ObjCEmojiPinner/ObjCEmojiPinner MessagesExtension/MessagesViewController.m
//
//  MessagesViewController.m
//  ObjCEmojiPinner MessagesExtension
//
//  Created by Ayush Satyavarpu on 9/16/23.
//

#import "MessagesViewController.h"
#import "CDHelper.h"
#import <QuartzCore/QuartzCore.h>

@implementation MessagesViewController

//COMPLETE
- (void)viewDidLoad {
    [super viewDidLoad];
    [self initializeVariables];
    [self loadData:YES];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillShow:)
                                                 name:UIKeyboardWillShowNotification
                                               object:nil];

    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(keyboardWillHide:)
                                                 name:UIKeyboardWillHideNotification
                                               object:nil];

    // Add a tap gesture recognizer
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self
                                                                                 action:@selector(handleTap)];
    [self.view addGestureRecognizer:tapGesture];
}

//COMPLETE
- (void) handleTap {
    [self.view endEditing:YES];
}

//COMPLETE
- (void) initializeTextField {
    [self.view addSubview:self.addButton];
    self.addButton.configuration = UIButtonConfiguration.filledButtonConfiguration;
    self.addButton.translatesAutoresizingMaskIntoConstraints = NO;
    if(self.addConstraint == nil){
        self.addConstraint = [self.addButton.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-10];
    }
    [NSLayoutConstraint activateConstraints:@[[self.addButton.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-10]]];
    self.addConstraint.active = YES;
    [self.addButton setTitle:@"Add" forState:UIControlStateNormal];
    [self.addButton addTarget:self action:@selector(addEmoji) forControlEvents:UIControlEventTouchUpInside];
    
    [self.view addSubview:self.mainTextField];
    if(self.mainTextConstraint == nil){
        self.mainTextConstraint = [self.mainTextField.bottomAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor constant:-10];
    }
    [NSLayoutConstraint activateConstraints:@[
        [self.mainTextField.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:10],
        [self.mainTextField.trailingAnchor constraintEqualToAnchor:self.addButton.leadingAnchor constant:-5],
        [self.mainTextField.topAnchor constraintEqualToAnchor:self.addButton.topAnchor]
    ]];
    
    self.mainTextConstraint.active = true;
    self.mainTextField.translatesAutoresizingMaskIntoConstraints = NO;
    self.mainTextField.placeholder = @"Enter Emoji";
    self.mainTextField.font = [UIFont systemFontOfSize:15];
    self.mainTextField.borderStyle = UITextBorderStyleRoundedRect;
    self.mainTextField.keyboardType = UIKeyboardTypeDefault;
    self.mainTextField.backgroundColor = UIColor.systemGray4Color;
}

//COMPLETED
- (void) keyboardWillShow:(NSNotification *) notification {
    NSDictionary * userInfo = notification.userInfo;
    CGRect keyboardFrame = [[userInfo objectForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
    
    CGFloat kbHeight = keyboardFrame.size.height;
    
    [UIView animateWithDuration:0.2 animations:^{
        self.mainTextConstraint.constant = -kbHeight + 30;
        self.addConstraint.constant = -kbHeight + 30;
        [self.view layoutIfNeeded];
    }];
}

//COMPLETED
- (void) keyboardWillHide:(NSNotification *) notification {
    [UIView animateWithDuration:0.2 animations:^{
        self.mainTextConstraint.constant = -10;
        self.addConstraint.constant = -10;
        [self.view layoutIfNeeded];
    }];
}

//COMPLETE
- (void) addEmoji {
    unichar insertChar = [self.mainTextField.text characterAtIndex:0];
    if(insertChar){
        [self writeData: [NSString stringWithCharacters:&insertChar length:1]];
        [self loadData:YES];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.mainTextField resignFirstResponder];
            self.mainTextField.text = @"";
            [self clearXButtons];
            [self addXButtons];
        });
    }else{
        [self.mainTextField resignFirstResponder];
    }
}

//COMPLETE
- (void) setupLabel {
    [self.view addSubview:self.label];
    self.label.text = @"F A V O R I T E D   E M O J I S";
    self.label.font = [UIFont systemFontOfSize:23 weight:UIFontWeightBold];
    self.label.textColor = UIColor.systemGray2Color;
    self.label.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [self.label.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:10.0],
        [self.label.topAnchor constraintEqualToAnchor:self.view.topAnchor constant:5.0]
    ]];
}

//COMPLETE
- (void) writeData:(NSString *)newEmojiCharacter {
    EmojiEntity * newEmoji = [[EmojiEntity alloc] initWithContext:self.context];
    newEmoji.emojiString = newEmojiCharacter;
    NSError *saveError = nil;
    if(![self.context save:&saveError]){
        NSLog(@"Error saving context: %@", saveError);
    }
    [self loadData:NO];
}

//COMPLETE
- (CGPoint)getPosition:(NSInteger)index {
    NSInteger rowIndex = floor((float)index / (float)self.numColumns);
    NSInteger colIndex = index % self.numColumns;
    
    CGFloat xOffset = (colIndex + 1) * (self.view.frame.size.width - 30) / self.numColumns;
    CGFloat yOffset = rowIndex * 40 + 60;
    
    return CGPointMake(xOffset, yOffset);
}

//COMPLETE
- (void) setupButtons {
    if(self.emojis.count == 0){
        [self.view addSubview:self.noEmojisLabel];
        self.noEmojisLabel.text = @"No emojis favorited yet. Swipe up to add some!";
        self.noEmojisLabel.font = [UIFont systemFontOfSize:15 weight:UIFontWeightBold];
        self.noEmojisLabel.textColor = UIColor.systemGray2Color;
        [NSLayoutConstraint activateConstraints:@[
            [self.noEmojisLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:10],
            [self.noEmojisLabel.topAnchor constraintEqualToAnchor:self.label.bottomAnchor constant:10]
        ]];
        self.noEmojisLabel.translatesAutoresizingMaskIntoConstraints = NO;
        return;
    }
    NSUInteger count = self.emojis.count;
    for(int i = 0;i<count;i++){
        EmojiEntity * emoji = self.emojis[i];
        UIButton * currButton = [[UIButton alloc] init];
        [self.view addSubview:currButton];
        currButton.configuration = UIButtonConfiguration.plainButtonConfiguration;
        currButton.translatesAutoresizingMaskIntoConstraints = NO;
        CGPoint pt = [self getPosition:i];
        CGFloat xOffset = pt.x;
        CGFloat yOffset = pt.y;
        NSLayoutConstraint * newTrailingConstraint = [currButton.trailingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:xOffset];
        NSLayoutConstraint * newCenterConstraint = [currButton.centerYAnchor constraintEqualToAnchor:self.view.topAnchor constant:yOffset];
        [NSLayoutConstraint activateConstraints:@[newTrailingConstraint, newCenterConstraint]];
        [currButton setTitle:emoji.emojiString forState:UIControlStateNormal];
        currButton.titleLabel.font = [UIFont systemFontOfSize:self.emojiFontSize];
        [currButton addTarget:self action:@selector(enterEmoji:) forControlEvents:UIControlEventTouchUpInside];
        [currButton addTarget:self action:@selector(onUp:) forControlEvents:UIControlEventTouchDown];
        [currButton.configuration setContentInsets: NSDirectionalEdgeInsetsMake(0.0, 0.0, 0.0, 0.0)];
        [self.emojiButtons addObject: [[UIButtonWrapper alloc] initWith:currButton :newTrailingConstraint :newCenterConstraint]];
    }
}

//COMPLETE
- (void) updateUI {
    for(UIButtonWrapper * button in self.emojiButtons){
        [button.button removeFromSuperview];
    }
    [self.noEmojisLabel removeFromSuperview];
    [self.emojiButtons removeAllObjects];
    
    [self setupLabel];
    [self setupButtons];
}

//COMPLETE
- (void) loadData: (BOOL) updateUI{
    NSFetchRequest * fr = [EmojiEntity fetchRequest];
    NSError * fetchError = nil;
    self.emojis = [[self.context executeFetchRequest:fr error:&fetchError] mutableCopy];
    
    if(fetchError){
        NSLog(@"Error loading data");
        return;
    }
    
    if(updateUI){
        dispatch_async(dispatch_get_main_queue(), ^{
            [self updateUI];
        });
    }
}

//COMPLETE
- (void) enterEmoji: (UIButton *) btn {
    if(self.isExpanded){
        return;
    }
    [UIView animateWithDuration:0.15 animations:^{
        btn.transform = CGAffineTransformIdentity;
    }];
    NSString * title = [btn titleForState:UIControlStateNormal];
    
    void (^myCompletionHandler)(NSError *error) = ^(NSError *error) {
        if (error) {
            NSLog(@"Error: %@", error.localizedDescription);
        } else {
            NSLog(@"No error occurred.");
        }
    };
    
    if(title) {
        [self.activeConversation insertText:title completionHandler:myCompletionHandler];
    }
}

//COMPLETE
- (void) onUp: (UIButton *) btn {
    if(self.isExpanded){
        return;
    }
    [UIView animateWithDuration:0.15 animations:^{
        btn.transform = CGAffineTransformMakeScale(0.8, 0.8);
    }];
    
}

//COMPLETE
- (void) initializeVariables {
    self.emojis = [[NSMutableArray<EmojiEntity*> array] init];
    self.context = CDHelper.persistentContainer.viewContext;
    self.emojiButtons = [[NSMutableArray<UIButtonWrapper*> array] init];
    self.xButtons = [[NSMutableArray<UIButton*> array] init];
    self.label = [[UILabel alloc] init];
    self.noEmojisLabel = [[UILabel alloc] init];
    self.isExpanded = NO;
    self.mainTextField = [[UITextField alloc] init];
    self.mainTextConstraint = nil;
    self.addButton = [[UIButton alloc] init];
    self.addConstraint = nil;
    self.numColumns = 7;
    self.emojiFontSize = 30;
}

//COMPLETE
- (void) removeEmoji:(UIButton *) sender{
    EmojiEntity * emojiToDelete = self.emojis[sender.tag];
    [self.context deleteObject:emojiToDelete];
    NSError *saveError = nil;
    if(![self.context save:&saveError]){
        NSLog(@"Error saving context: %@", saveError);
    }
    [self loadData:NO];
    UIButtonWrapper * emojiButtonToDelete = self.emojiButtons[sender.tag];
    UIButton * xButtonToDelete = self.xButtons[sender.tag];
    [self.emojiButtons removeObjectAtIndex:sender.tag];
    [self.xButtons removeObjectAtIndex:sender.tag];
    for(UIButton * xButton in self.xButtons){
        if(xButton.tag > sender.tag){
            xButton.tag = xButton.tag - 1;
        }
    }
    [UIView animateWithDuration:0.2 animations:^{
        for(NSUInteger i = sender.tag; i<self.emojiButtons.count;i++){
            UIButtonWrapper * emojiButton = self.emojiButtons[i];
            CGPoint pt = [self getPosition:i];
            emojiButton.centerYConstraint.constant = pt.y;
            emojiButton.trailingConstraint.constant = pt.x;
        }
        [self.view layoutIfNeeded];
        emojiButtonToDelete.button.alpha = 0.0;
        xButtonToDelete.alpha = 0.0;
    } completion:^(BOOL completed){
        [emojiButtonToDelete.button removeFromSuperview];
        [xButtonToDelete removeFromSuperview];
    }];
    
}

//COMPLETE
- (void) addXButtons {
    NSUInteger count = [self.emojiButtons count];
    for(NSUInteger i = 0;i<count;i++){
        UIButtonWrapper * emojiButton = self.emojiButtons[i];
        [self shakeButton:emojiButton.button];
        UIButton * newX = [[UIButton alloc] init];
        [self.view addSubview:newX];
        newX.configuration = UIButtonConfiguration.grayButtonConfiguration;
        newX.tintColor = UIColor.whiteColor;
        newX.translatesAutoresizingMaskIntoConstraints = NO;
        [NSLayoutConstraint activateConstraints:@[
            [newX.centerXAnchor constraintEqualToAnchor:emojiButton.button.leadingAnchor constant:10.0],
            [newX.centerYAnchor constraintEqualToAnchor:emojiButton.button.topAnchor constant:10.0]
        ]];
        [newX setTitle:@"x" forState:UIControlStateNormal];
        newX.titleLabel.font = [UIFont systemFontOfSize:10];
        newX.tag = i;
        [newX addTarget:self action:@selector(removeEmoji:) forControlEvents:UIControlEventTouchUpInside];
        [newX.configuration setContentInsets: NSDirectionalEdgeInsetsMake(0, 200.0, 200.0, 2.0)];
        [self.view bringSubviewToFront:newX];
        [self.xButtons addObject:newX];
        [self shakeButtonViolently:newX];
    }
}

//COMPLETE
- (void) clearXButtons {
    for(UIButton * btn in self.xButtons){
        [btn removeFromSuperview];
    }
    self.xButtons = [NSMutableArray<UIButton*> array];
}

//COMPLETE
- (void)shakeButton:(UIButton *) button {
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    animation.duration = 0.1;  // Adjust the duration as needed
    animation.autoreverses = YES;
    animation.repeatCount = HUGE_VALF;  // Repeat indefinitely
    animation.fromValue = @(M_PI / -18.0); // Rotate slightly counterclockwise (about 0.05 radians)
    animation.toValue = @(M_PI / 18.0);   // Rotate slightly clockwise (about 0.05 radians)
    [button.layer addAnimation:animation forKey:@"rotation"];
}

//COMPLETE
- (void)shakeButtonViolently:(UIButton *) button {
    CABasicAnimation *animation = [CABasicAnimation animationWithKeyPath:@"transform.rotation.z"];
    animation.duration = 0.1;
    animation.autoreverses = YES;
    animation.repeatCount = HUGE_VALF;
    animation.fromValue = @(M_PI / -45.0);
    animation.toValue = @(M_PI / 45.0);
    [button.layer addAnimation:animation forKey:@"rotationViolent"];
}

//COMPLETE
-(void)willTransitionToPresentationStyle:(MSMessagesAppPresentationStyle)presentationStyle {
    // Called before the extension transitions to a new presentation style.
    if(presentationStyle == MSMessagesAppPresentationStyleExpanded){
        [self addXButtons];
        self.isExpanded = YES;
        [self initializeTextField];
        self.mainTextConstraint.constant = -10;
        self.addConstraint.constant = -10;
        [self.view layoutIfNeeded];
    }else if(presentationStyle == MSMessagesAppPresentationStyleCompact){
        for(UIButtonWrapper * emojiButton in self.emojiButtons){
            [emojiButton.button.layer removeAnimationForKey:@"rotation"];
        }
        [self clearXButtons];
        self.isExpanded = NO;
        [self loadData:YES];
        self.mainTextConstraint.constant = -10;
        self.addConstraint.constant = -10;
        [self.view layoutIfNeeded];
    }
}

@end

File: EmojiPinner/EmojiPinner MessagesExtension/MessagesViewController.swift
//
//  MessagesViewController.swift
//  PinnedEmojis MessagesExtension
//
//  Created by Ayush Satyavarpu on 9/4/23.
//

import UIKit
import Messages
import CoreData

func generateRandomEmoji() -> String {
    let emojiRanges: [ClosedRange<UInt32>] = [
        0x1F600...0x1F64F
    ]

    let randomRange = emojiRanges.randomElement()!
    let randomEmoji = UnicodeScalar(randomRange.lowerBound + arc4random_uniform(randomRange.upperBound - randomRange.lowerBound + 1))!
    return String(randomEmoji)
}

struct UIButtonWrapper{
    var button: UIButton
    var trailingConstraint: NSLayoutConstraint
    var centerYConstraint: NSLayoutConstraint
}

class MessagesViewController: MSMessagesAppViewController {
            
    var emojis: [EmojiEntity] = []
    var context = CoreDataHelper.persistentContainer.viewContext
    var emojiButtons: [UIButtonWrapper] = []
    var xButtons: [UIButton] = []
    var label = UILabel()
    var noEmojisLabel = UILabel()
    var isExpanded = false
    var mainTextField = UITextField()
    var mainTextConstraint:NSLayoutConstraint? = nil
    var addButton = UIButton()
    var addConstraint:NSLayoutConstraint? = nil
    
    let numColumns = 7
    let emojiFontSize: CGFloat = 30
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        loadData(true)
        updateUI()
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(keyboardWillShow(_:)),
            name: UIResponder.keyboardWillShowNotification,
            object: nil
        )
        
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(keyboardWillHide(_:)),
            name: UIResponder.keyboardWillHideNotification,
            object: nil
        )
        
        let tapGesture = UITapGestureRecognizer(target: self, action: #selector(handleTap))
        view.addGestureRecognizer(tapGesture)
    }
    
    @objc func handleTap(){
        view.endEditing(true)
    }
    
    func initializeTextField(){
        
        self.view.addSubview(addButton)
        addButton.configuration = .filled()
        addButton.translatesAutoresizingMaskIntoConstraints = false
        if addConstraint == nil{
            addConstraint = addButton.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor, constant: -10)
        }
        NSLayoutConstraint.activate([
            addButton.trailingAnchor.constraint(equalTo: self.view.trailingAnchor, constant: -10),
        ])
        self.addConstraint?.isActive = true
        self.addButton.configuration?.title = "Add"
        self.addButton.addTarget(self, action: #selector(addEmoji), for: .touchUpInside)

        
        self.view.addSubview(mainTextField)
        if self.mainTextConstraint == nil{
            self.mainTextConstraint = mainTextField.bottomAnchor.constraint(equalTo: self.view.safeAreaLayoutGuide.bottomAnchor, constant: -10)
        }
        NSLayoutConstraint.activate([
            mainTextField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
            mainTextField.trailingAnchor.constraint(equalTo: self.addButton.leadingAnchor, constant: -5),
            mainTextField.topAnchor.constraint(equalTo: self.addButton.topAnchor)
        ])
        self.mainTextConstraint?.isActive = true
        mainTextField.translatesAutoresizingMaskIntoConstraints = false
        mainTextField.placeholder = "Enter emoji"
        mainTextField.font = UIFont.systemFont(ofSize: 15)
        mainTextField.borderStyle = .roundedRect
        mainTextField.keyboardType = .default
        mainTextField.backgroundColor = UIColor.systemGray4
        
        

        
    }
    
    @objc func keyboardWillShow(_ notification : Notification){
        print("Keyboard will show")
        guard let userInfo = notification.userInfo, let keyboardFrame = userInfo[UIResponder.keyboardFrameEndUserInfoKey] as? CGRect else{
            return
        }
        let kbHeight = keyboardFrame.height
        UIView.animate(withDuration: 0.2, animations: {
            self.mainTextConstraint?.constant = -kbHeight + 30
            self.addConstraint?.constant = -kbHeight + 30
            self.view.layoutIfNeeded()
        })
    }
    
    @objc func keyboardWillHide(_ notification: Notification){
        UIView.animate(withDuration: 0.2, animations: {
            self.mainTextConstraint?.constant = -10
            self.addConstraint?.constant = -10
            self.view.layoutIfNeeded()
        })
    }
    
    @objc func removeEmojis(){
        do{
            self.emojis = try self.context.fetch(EmojiEntity.fetchRequest())
            for e in self.emojis{
                context.delete(e)
            }
            do{
                try self.context.save()
            }catch {
                print("error in wiping data \(error)")
            }
            loadData(true)
        }catch{
            print("error in loadData \(error)")
        }
    }
    
    @objc func removeEmoji(_ sender: UIButton){
        print("removing \(sender.tag)")
        let emojiToDelete = self.emojis[sender.tag]
        context.delete(emojiToDelete)
        do{
            try self.context.save()
        }catch {
            print("error in wiping data \(error)")
        }
        loadData(false)
        let emojiButtonToDelete = self.emojiButtons[sender.tag]
        let xButtonToDelete = self.xButtons[sender.tag]
        self.emojiButtons.remove(at: sender.tag)
        self.xButtons.remove(at: sender.tag)
        for xButton in xButtons{
            if xButton.tag > sender.tag{
                xButton.tag = xButton.tag - 1
            }
        }
        UIView.animate(withDuration: 0.2, animations: {
            for i in sender.tag..<self.emojiButtons.count {
                    let emojiButton = self.emojiButtons[i]
                    let (xOffset, yOffset): (CGFloat, CGFloat) = self.getPosition(i)
                    
                    // Update the constant values of existing constraints
                    emojiButton.centerYConstraint.constant = yOffset
                    emojiButton.trailingConstraint.constant = xOffset
                }
                
                // Layout the views with animations
                self.view.layoutIfNeeded()
                
                // Fade out and remove the deleted button
                emojiButtonToDelete.button.alpha = 0.0
                xButtonToDelete.alpha = 0.0
            }) { (completed) in
                // After the animation is complete, remove the deleted button from the view hierarchy
                emojiButtonToDelete.button.removeFromSuperview()
                xButtonToDelete.removeFromSuperview()
            }
    }
    
    @objc func addEmoji(){
        
        if let insertChar: Character = mainTextField.text!.first {
            writeData(String(insertChar))
            self.loadData(true)
            DispatchQueue.main.async{
                self.mainTextField.resignFirstResponder()
                self.mainTextField.text = ""
                self.removeXButtons()
                self.addXButtons()
            }
            
            
        }else {
            mainTextField.resignFirstResponder()
        }
        
    }
    
    @objc func writeData(_ newEmojiCharacter: String){
        let newEmoji = EmojiEntity(context: self.context)
        newEmoji.emojiString = newEmojiCharacter
        do{
            try self.context.save()
        }catch{
            print("error in writeData \(error)")
        }
        loadData(false)
    }
    
    
    
    func loadData(_ updateUI: Bool){
        do{
            self.emojis = try self.context.fetch(EmojiEntity.fetchRequest())
            if updateUI{
                DispatchQueue.main.async{
                    self.updateUI()
                }
            }
        }catch{
            print("error in loadData \(error)")
        }
        
    }
    
    func setupLabel(){
        view.addSubview(label)
        label.text = "F A V O R I T E D   E M O J I S"
        label.font = UIFont.systemFont(ofSize: 23, weight: .bold)
        label.textColor = .systemGray2
        label.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            label.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
            label.topAnchor.constraint(equalTo: view.topAnchor, constant: 5)
        ])
    }
    
    func updateUI() {
        // Remove existing buttons from the view
        for button in emojiButtons {
            button.button.removeFromSuperview()
        }
        noEmojisLabel.removeFromSuperview()
        emojiButtons.removeAll()

        // Re-create buttons based on the updated emojis array
        setupLabel()
        setupButtons()
    }
    
    func getPosition(_ index: Int) -> (CGFloat, CGFloat){
        let rowIndex = floor(Float(Float(index)/Float(numColumns)))
        let colIndex = index % numColumns
        
        let xOffset = Float(colIndex + 1) * Float(self.view.frame.width - 30) / Float(numColumns)
        let yOffset = rowIndex * 40 + 60
        
        return (CGFloat(xOffset), CGFloat(yOffset))
    }
    
    func setupButtons(){
        if(emojis.isEmpty){
            view.addSubview(noEmojisLabel)
            noEmojisLabel.text = "No emojis favorited yet. Swipe up to add some!"
            noEmojisLabel.font = UIFont.systemFont(ofSize: 15, weight: .bold)
            noEmojisLabel.textColor = .systemGray2
            NSLayoutConstraint.activate([
                noEmojisLabel.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 10),
                noEmojisLabel.topAnchor.constraint(equalTo: label.bottomAnchor, constant: 10)
            ])
            noEmojisLabel.translatesAutoresizingMaskIntoConstraints = false
            return
        }
        for (i, emoji) in emojis.enumerated() {
            let currButton = UIButton()
            view.addSubview(currButton)
            currButton.configuration = .plain()
            currButton.translatesAutoresizingMaskIntoConstraints = false
            let (xOffset, yOffset) : (CGFloat, CGFloat) = getPosition(i)
            let newTrailingConstraint = currButton.trailingAnchor.constraint(equalTo: self.view.leadingAnchor, constant: xOffset)
            let newCenterConstraint = currButton.centerYAnchor.constraint(equalTo: self.view.topAnchor, constant: yOffset)
            NSLayoutConstraint.activate([newTrailingConstraint, newCenterConstraint])
            currButton.configuration?.title = emoji.emojiString
            currButton.configuration?.attributedTitle = AttributedString(currButton.configuration!.title!, attributes: AttributeContainer([NSAttributedString.Key.font : UIFont.systemFont(ofSize: emojiFontSize)]))
            currButton.addTarget(self, action: #selector(enterEmoji(_:)), for: .touchUpInside)
            currButton.addTarget(self, action:#selector(onUp(_:)), for: .touchDown)
            currButton.configuration!.contentInsets = NSDirectionalEdgeInsets(top: 0.0, leading: 0.0, bottom: 0.0, trailing: 0.0)
            emojiButtons.append(UIButtonWrapper(button: currButton, trailingConstraint: newTrailingConstraint, centerYConstraint: newCenterConstraint))
        }


    }
    
    @objc func onUp(_ btn: UIButton){
        if isExpanded{
            return
        }
        UIView.animate(withDuration: 0.15) {
                btn.transform = CGAffineTransform(scaleX: 0.8, y: 0.8)
            }
    }
    
    @objc func enterEmoji(_ btn: UIButton) {
        if isExpanded{
            return
        }
        UIView.animate(withDuration: 0.15) {
                btn.transform = CGAffineTransform.identity
            }
        if let title = btn.configuration?.title {
            self.activeConversation?.insertText(title)
        }
    }
    
    // MARK: - Conversation Handling
    
    override func willBecomeActive(with conversation: MSConversation) {
        // Called when the extension is about to move from the inactive to active state.
        // This will happen when the extension is about to present UI.
        
        // Use this method to configure the extension and restore previously stored state.
    }
    
    override func didResignActive(with conversation: MSConversation) {
        // Called when the extension is about to move from the active to inactive state.
        // This will happen when the user dismisses the extension, changes to a different
        // conversation or quits Messages.
        
        // Use this method to release shared resources, save user data, invalidate timers,
        // and store enough state information to restore your extension to its current state
        // in case it is terminated later.
    }
   
    override func didReceive(_ message: MSMessage, conversation: MSConversation) {
        // Called when a message arrives that was generated by another instance of this
        // extension on a remote device.
        
        // Use this method to trigger UI updates in response to the message.
    }
    
    override func didStartSending(_ message: MSMessage, conversation: MSConversation) {
        // Called when the user taps the send button.
    }
    
    override func didCancelSending(_ message: MSMessage, conversation: MSConversation) {
        // Called when the user deletes the message without sending it.
    
        // Use this to clean up state related to the deleted message.
    }
    
    func addXButtons(){
        for (i, emojiButton) in emojiButtons.enumerated() {
            shakeButton(button: emojiButton.button)
            let newX = UIButton()
            view.addSubview(newX)
            newX.configuration = .gray()
            newX.tintColor = .white
            newX.translatesAutoresizingMaskIntoConstraints = false
            NSLayoutConstraint.activate([
                newX.centerXAnchor.constraint(equalTo: emojiButton.button.leadingAnchor, constant: 10.0),
                newX.centerYAnchor.constraint(equalTo: emojiButton.button.topAnchor, constant: 10.0)
            ])
            newX.configuration?.title = "x"
            newX.configuration?.attributedTitle = AttributedString(newX.configuration!.title!, attributes: AttributeContainer([NSAttributedString.Key.font : UIFont.systemFont(ofSize: 10)]))
            newX.tag = i
            newX.addTarget(self, action: #selector(removeEmoji), for: .touchUpInside)
            newX.configuration!.contentInsets = NSDirectionalEdgeInsets(top: 0, leading: 2, bottom: 2, trailing: 2)
            view.bringSubviewToFront(newX)
            xButtons.append(newX)
            shakeButtonViolently(button: newX)
        }
    }
    
    func removeXButtons(){
        for xButton in xButtons{
            xButton.removeFromSuperview()
        }
        xButtons = []
    }
    
    func shakeButton(button: UIButton) {
        let animation = CABasicAnimation(keyPath: "transform.rotation.z")
        animation.duration = 0.1  // Adjust the duration as needed
        animation.autoreverses = true
        animation.repeatCount = .infinity  // Repeat indefinitely
        animation.fromValue = NSNumber(value: Float(-0.05)) // Rotate slightly counterclockwise
        animation.toValue = NSNumber(value: Float(0.05))   // Rotate slightly clockwise
        button.layer.add(animation, forKey: "rotation")
    }
    
    
    func shakeButtonViolently(button: UIButton) {
        let animation = CABasicAnimation(keyPath: "transform.rotation.z")
        animation.duration = 0.1  // Adjust the duration as needed
        animation.autoreverses = true
        animation.repeatCount = .infinity  // Repeat indefinitely
        animation.fromValue = NSNumber(value: Float(-0.13)) // Rotate slightly counterclockwise
        animation.toValue = NSNumber(value: Float(0.13))   // Rotate slightly clockwise
        button.layer.add(animation, forKey: "rotationViolent")
    }
    
    override func willTransition(to presentationStyle: MSMessagesAppPresentationStyle) {
        if presentationStyle == .expanded{
            addXButtons()
            self.isExpanded = true
            self.initializeTextField()
            self.mainTextConstraint?.constant = -10
            self.addConstraint?.constant = -10
            self.view.layoutIfNeeded()
        }else if presentationStyle == .compact{
            for emojiButton in emojiButtons {
                emojiButton.button.layer.removeAnimation(forKey: "rotation")
            }
            removeXButtons()
            self.isExpanded = false
            self.loadData(true)
            self.mainTextConstraint?.constant = -10
            self.addConstraint?.constant = -10
            self.view.layoutIfNeeded()
        }
    }
    
    override func didTransition(to presentationStyle: MSMessagesAppPresentationStyle) {
        // Called after the extension transitions to a new presentation style.se this method to finalize any behaviors associated with the change in presentation style.
    }

}
